monitoring/v1/types.go

```diff
@@ -0,0 +1,337 @@
+// +openshift:compatibility-gen:level=1
+// +k8s:openapi-gen=true
+type AlertRelabelConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the desired state of this AlertRelabelConfig object.
+	Spec AlertRelabelConfigSpec `json:"spec"`
+
+	// status describes the current state of this AlertRelabelConfig object.
+	//
+	// +optional
+	Status AlertRelabelConfigStatus `json:"status,omitempty"`
+}
+
+// AlertRelabelConfigsSpec is the desired state of an AlertRelabelConfig resource.
+//
+// +k8s:openapi-gen=true
+type AlertRelabelConfigSpec struct {
+	// configs is a list of sequentially evaluated alert relabel configs.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Configs []RelabelConfig `json:"configs"`
+}
+
+// AlertRelabelConfigStatus is the status of an AlertRelabelConfig resource.
+type AlertRelabelConfigStatus struct {
+	// conditions contains details on the state of the AlertRelabelConfig, may be
+	// empty.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+const (
+	// AlertRelabelConfigReady is the condition type indicating readiness.
+	AlertRelabelConfigReady string = "Ready"
+)
+
+// AlertRelabelConfigList is a list of AlertRelabelConfigs.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type AlertRelabelConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// items is a list of AlertRelabelConfigs.
+	Items []*AlertRelabelConfig `json:"items"`
+}
+
+// LabelName is a valid Prometheus label name which may only contain ASCII
+// letters, numbers, and underscores.
+//
+// +kubebuilder:validation:Pattern:="^[a-zA-Z_][a-zA-Z0-9_]*$"
+type LabelName string
+
+// RelabelConfig allows dynamic rewriting of label sets for alerts.
+// See Prometheus documentation:
+// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#alert_relabel_configs
+// - https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config
+//
+// +k8s:openapi-gen=true
+type RelabelConfig struct {
+	// sourceLabels select values from existing labels. Their content is
+	// concatenated using the configured separator and matched against the
+	// configured regular expression for the Replace, Keep, and Drop actions.
+	//
+	// +optional
+	SourceLabels []LabelName `json:"sourceLabels,omitempty"`
+
+	// separator placed between concatenated source label values. When omitted,
+	// Prometheus will use its default value of ';'.
+	//
+	// +optional
+	Separator string `json:"separator,omitempty"`
+
+	// targetLabel to which the resulting value is written in a 'Replace' action.
+	// It is mandatory for 'Replace' and 'HashMod' actions. Regex capture groups
```

Comment:
Since this field is required when action has a certain value, we could use a CEL validation to make sure this is set, add this to the parent struct
```
// +kubebuilder:validation:XValidation:rule="self.action != 'Replace' && self.action != 'HashMod' || has(self.targetLabel)",message="targetLabel is required when action is Replace or HashMod"
```
The logic here is either action isn't HashMod/Replace, or we have the targetLabel